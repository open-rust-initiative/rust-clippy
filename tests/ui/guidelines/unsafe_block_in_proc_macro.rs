#![crate_type = "proc-macro"]
#![warn(clippy::unsafe_block_in_proc_macro)]
#![allow(clippy::needless_return)]

extern crate proc_macro;
extern crate quote;

use proc_macro::TokenStream;
use quote::quote;

fn do_something() {}
fn main() {}

#[proc_macro]
pub fn unsafe_print_foo(_: TokenStream) -> TokenStream {
    quote!({
        unsafe {
            println!("foo");
        }
    })
    .into()
}

#[rustfmt::skip]
#[proc_macro]
pub fn unsafe_print_bar_unformatted(_: TokenStream) -> TokenStream {
    quote!({
        unsafe
        
        
        { println!("bar"); }
    }).into()
}

#[rustfmt::skip]
#[proc_macro]
pub fn unsafe_print_bar_unformatted_1(_: TokenStream) -> TokenStream {
    quote!({
        unsafe{println!("bar");}
    }).into()
}

#[proc_macro]
pub fn unsafe_print_baz(_: TokenStream) -> TokenStream {
    do_something();

    quote!({
        let _ = 0_u8;
        do_something();
        unsafe {
            println!("baz");
        }

        do_something();
    })
    .into()
}

#[proc_macro]
pub fn maybe_unsafe_print(_: TokenStream) -> TokenStream {
    do_something();

    if false {
        return quote!({
            unsafe {
                println!("unsafe");
            }
        })
        .into();
    }

    quote!({
        println!("safe");
    })
    .into()
}

#[proc_macro]
pub fn maybe_unsafe_print_1(_: TokenStream) -> TokenStream {
    let condition = 1;
    if condition > 0 {
        return quote!({
            println!("safe");
        })
        .into();
    }

    quote!({
        unsafe {
            println!("unsafe");
        }
    })
    .into()
}

#[proc_macro]
pub fn maybe_unsafe_print_2(_: TokenStream) -> TokenStream {
    let condition = 1;
    if condition == 0 {
        return quote!({}).into();
    } else if condition == 1 {
        return quote!(unsafe {}).into();
    } else if condition == 2 {
        return quote!({ println!("2") }).into();
    }
    do_something();
    quote!({}).into()
}

#[proc_macro]
pub fn multiple_unsafe(_: TokenStream) -> TokenStream {
    let condition = 1;
    match condition {
        1 => quote!(unsafe {
            println!("1");
        }),
        2 => quote!(unsafe {
            println!("2");
        }),
        _ => quote!(unsafe {}),
    }
    .into()
}

#[proc_macro]
pub fn unsafe_block_in_function(_: TokenStream) -> TokenStream {
    quote!({
        fn print_foo() {
            unsafe {
                println!("foo");
            }
        }
    })
    .into()
}

#[proc_macro]
// Don't lint
pub fn print_foo(_: TokenStream) -> TokenStream {
    quote!({
        println!("foo");
    })
    .into()
}

#[proc_macro]
// Don't lint
pub fn unsafe_trait(_: TokenStream) -> TokenStream {
    quote!({
        unsafe trait Foo {
            fn aaa();
            fn bbb();
        }
    })
    .into()
}

#[proc_macro]
// Don't lint
pub fn unsafe_fn(_: TokenStream) -> TokenStream {
    quote!({
        unsafe fn uns_foo() {}
    })
    .into()
}

#[proc_macro]
// Don't lint
pub fn unsafe_impl_fn(_: TokenStream) -> TokenStream {
    quote!({
        struct T;
        impl T {
            unsafe fn foo() {}
        }
    })
    .into()
}
