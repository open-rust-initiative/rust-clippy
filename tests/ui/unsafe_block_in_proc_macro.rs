#![warn(clippy::unsafe_block_in_proc_macro)]
#![allow(clippy::needless_return)]

extern crate proc_macro;
extern crate quote;

use proc_macro::TokenStream;
use quote::quote;

fn do_something() {}
fn main() {}

// Ideally we can add `#[proc_macro]` attr on these following functions,
// and check only such functions instead, but I don't know how to write such test
// without putting `proc_macro = true` in Cargo.toml, maybe someone else can fix it.

pub fn unsafe_print_foo(_: TokenStream) -> TokenStream {
    quote!({
        unsafe {
            println!("foo");
        }
    })
    .into()
}

#[rustfmt::skip]
pub fn unsafe_print_bar_unformatted(_: TokenStream) -> TokenStream {
    quote!({
        unsafe
        
        
        { println!("bar"); }
    }).into()
}

#[rustfmt::skip]
pub fn unsafe_print_bar_unformatted_1(_: TokenStream) -> TokenStream {
    quote!({
        unsafe{println!("bar");}
    }).into()
}

pub fn unsafe_print_baz(_: TokenStream) -> TokenStream {
    do_something();

    quote!({
        let _ = 0_u8;
        do_something();
        unsafe {
            println!("baz");
        }

        do_something();
    })
    .into()
}

pub fn maybe_unsafe_print(_: TokenStream) -> TokenStream {
    do_something();

    if false {
        return quote!({
            unsafe {
                println!("unsafe");
            }
        })
        .into();
    }

    quote!({
        println!("safe");
    })
    .into()
}

pub fn maybe_unsafe_print_1(_: TokenStream) -> TokenStream {
    let condition = 1;
    if condition > 0 {
        return quote!({
            println!("safe");
        })
        .into();
    }

    quote!({
        unsafe {
            println!("unsafe");
        }
    })
    .into()
}

pub fn maybe_unsafe_print_2(_: TokenStream) -> TokenStream {
    let condition = 1;
    if condition == 0 {
        return quote!({}).into();
    } else if condition == 1 {
        return quote!(unsafe {}).into();
    } else if condition == 2 {
        return quote!({ println!("2") }).into();
    }
    do_something();
    quote!({}).into()
}

pub fn multiple_unsafe(_: TokenStream) -> TokenStream {
    let condition = 1;
    match condition {
        1 => quote!(unsafe {
            println!("1");
        }),
        2 => quote!(unsafe {
            println!("2");
        }),
        _ => quote!(unsafe {}),
    }
    .into()
}

pub fn unsafe_block_in_function(_: TokenStream) -> TokenStream {
    quote!({
        fn print_foo() {
            unsafe {
                println!("foo");
            }
        }
    })
    .into()
}

// Don't lint
pub fn print_foo(_: TokenStream) -> TokenStream {
    quote!({
        println!("foo");
    })
    .into()
}

// Don't lint
pub fn unsafe_trait(_: TokenStream) -> TokenStream {
    quote!({
        unsafe trait Foo {
            fn aaa();
            fn bbb();
        }
    })
    .into()
}

// Don't lint
pub fn unsafe_fn(_: TokenStream) -> TokenStream {
    quote!({
        unsafe fn uns_foo() {}
    })
    .into()
}

// Don't lint
pub fn unsafe_impl_fn(_: TokenStream) -> TokenStream {
    quote!({
        struct T;
        impl T {
            unsafe fn foo() {}
        }
    })
    .into()
}
