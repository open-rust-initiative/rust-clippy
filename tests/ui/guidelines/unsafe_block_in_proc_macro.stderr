error: function-like procedural macro `unsafe_print_foo` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:16:5
   |
LL | /     quote!({
LL | |         unsafe {
LL | |             println!("foo");
LL | |         }
LL | |     })
   | |______^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           and add unsafe block when calling the macro instead: `unsafe { unsafe_print_foo!{...} }`
   = note: `-D clippy::unsafe-block-in-proc-macro` implied by `-D warnings`

error: function-like procedural macro `unsafe_print_bar_unformatted` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:27:5
   |
LL | /     quote!({
LL | |         unsafe
LL | |         
LL | |         
LL | |         { println!("bar"); }
LL | |     }).into()
   | |______^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           and add unsafe block when calling the macro instead: `unsafe { unsafe_print_bar_unformatted!{...} }`

error: function-like procedural macro `unsafe_print_bar_unformatted_1` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:38:5
   |
LL | /     quote!({
LL | |         unsafe{println!("bar");}
LL | |     }).into()
   | |______^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           and add unsafe block when calling the macro instead: `unsafe { unsafe_print_bar_unformatted_1!{...} }`

error: function-like procedural macro `unsafe_print_baz` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:47:5
   |
LL | /     quote!({
LL | |         let _ = 0_u8;
LL | |         do_something();
LL | |         unsafe {
...  |
LL | |         do_something();
LL | |     })
   | |______^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           and add unsafe block when calling the macro instead: `unsafe { unsafe_print_baz!{...} }`

error: function-like procedural macro `maybe_unsafe_print` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:64:16
   |
LL |           return quote!({
   |  ________________^
LL | |             unsafe {
LL | |                 println!("unsafe");
LL | |             }
LL | |         })
   | |__________^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           and add unsafe block when calling the macro instead: `unsafe { maybe_unsafe_print!{...} }`

error: function-like procedural macro `maybe_unsafe_print_1` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:88:5
   |
LL | /     quote!({
LL | |         unsafe {
LL | |             println!("unsafe");
LL | |         }
LL | |     })
   | |______^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           and add unsafe block when calling the macro instead: `unsafe { maybe_unsafe_print_1!{...} }`

error: function-like procedural macro `maybe_unsafe_print_2` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:102:16
   |
LL |         return quote!(unsafe {}).into();
   |                ^^^^^^^^^^^^^^^^^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           and add unsafe block when calling the macro instead: `unsafe { maybe_unsafe_print_2!{...} }`

error: function-like procedural macro `multiple_unsafe` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:114:14
   |
LL |           1 => quote!(unsafe {
   |  ______________^
LL | |             println!("1");
LL | |         }),
   | |__________^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           and add unsafe block when calling the macro instead: `unsafe { multiple_unsafe!{...} }`

error: function-like procedural macro `multiple_unsafe` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:117:14
   |
LL |           2 => quote!(unsafe {
   |  ______________^
LL | |             println!("2");
LL | |         }),
   | |__________^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           and add unsafe block when calling the macro instead: `unsafe { multiple_unsafe!{...} }`

error: function-like procedural macro `multiple_unsafe` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:120:14
   |
LL |         _ => quote!(unsafe {}),
   |              ^^^^^^^^^^^^^^^^^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           and add unsafe block when calling the macro instead: `unsafe { multiple_unsafe!{...} }`

error: function-like procedural macro `unsafe_block_in_function` could masking unsafe operations
  --> $DIR/unsafe_block_in_proc_macro.rs:127:5
   |
LL | /     quote!({
LL | |         fn print_foo() {
LL | |             unsafe {
LL | |                 println!("foo");
LL | |             }
LL | |         }
LL | |     })
   | |______^
   |
   = help: consider removing unsafe block from the above `quote!` call,
           if the block was in a function, put an `unsafe` keyword on its definition: `unsafe fn ...`, 
           otherwise add unsafe block when calling the macro instead: `unsafe { unsafe_block_in_function!{...} }`

error: aborting due to 11 previous errors

